crates/ty_python_semantic/src/types/cyclic.rs

```diff
@@ -1,23 +1,39 @@
 use crate::FxIndexSet;
 use crate::types::Type;
+use std::cmp::Eq;
+use std::hash::Hash;
 
-#[derive(Debug, Default)]
-pub(crate) struct TypeTransformer<'db> {
-    seen: FxIndexSet<Type<'db>>,
+pub(crate) type TypeTransformer<'db> = CycleDetector<Type<'db>, Type<'db>>;
+
+impl Default for TypeTransformer<'_> {
+    fn default() -> Self {
+        // TODO: proper recursive type handling
+
+        // This must be Any, not e.g. a todo type, because Any is the normalized form of the
+        // dynamic type (that is, todo types are normalized to Any).
+        CycleDetector::new(Type::any())
+    }
+}
+
+pub(crate) type PairVisitor<'db> = CycleDetector<(Type<'db>, Type<'db>), bool>;
+
+#[derive(Debug)]
+pub(crate) struct CycleDetector<T: Hash + Eq, R: Copy> {
```

Comment:
you can provide the bounds on the `impl` block without specifying them for the struct itself -- it's less noisy in terms of the syntax and also makes the struct definition more flexible (it would allow us to add `impl` blocks for instances of `CycleDetector` where `T: !Hash`, for example, if there was ever a need for those instances).

```suggestion
pub(crate) struct CycleDetector<T, R> {
```